url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+                        exception = null;
+                        // Awake the debugger.
+                        EvalTrace.this.notifyAll();
+                        // The eval must wait, otherwise it might attempt
+                        // to evaluate a brand new expression which will destroy the result of
+                        // this earlier evaluation!
+                        EvalTrace.this.wait();
+                        halted = false;
+                    }
+                    catch (InterruptedException e)
+                    {
+                        halted = true;
+                    }
+                }
+            }
+
+            public void startEval(Eval2.EvalEvent aEvent)
+            {
+                synchronized(EvalTrace.this)
+                {
+                    try
+                    {
+                        // The eval marks itself as being halted (see a bit further).
+                        halted = true;
+                        result = null;
+                        hasResult = false;
+                        excepted = false;
+                        exception = null;
+                        // The eval makes its stack available for the debugger.
+                        stack = aEvent.getStack();
+                        // It wakes up the debugger.
+                        EvalTrace.this.notifyAll();
+                        // The eval puts itself in wait on te trace manager.
+                        EvalTrace.this.wait();
+                        // If we pass through the wait method, the eval kicks into action.
+                        halted = false;
+                    }
+                    catch (InterruptedException e)
+                    {
+                        halted = true;
+                    }
+                }
+            }
+
+            public void stepEvent(Eval2.EvalEvent aEvent)
+            {
+                synchronized (EvalTrace.this)
+                {
+                    try
+                    {
+                        // Set the halted flag.
+                        halted = true;
+                        // Provide the stack for the debugger to examine.
+                        stack = aEvent.getStack();
+                        // Wake up the debugger.
+                        EvalTrace.this.notifyAll();
+                        // If there are more steps to take, we let the eval wait a bit.
+                        // If this was the last step, we can let the eval worker finish its job
+                        // and prepare the result for us.
+                        if(stack.hasMoreSteps())
+                        {
+                            // The eval goes to sleep ...
+                            // We wait for a signal of the tracer.
+                            EvalTrace.this.wait();
+                            halted = false;
+                        }
+                    }
+                    catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)